paasta_tools/automatic_rollbacks/slack.py

```diff
@@ -242,3 +91,234 @@ def scribe_tail(queue):
                 yield event
         except Empty:
             pass
+
+
+class SlackDeploymentProcess(DeploymentProcess, abc.ABC):
+    def __init__(self) -> None:
+        super().__init__()
+        self.human_readable_status = "Initializing..."
+        self.slack_client = self.get_slack_client()
+        self.last_action = None
+
+        slack_thread = Thread(target=self.listen_for_slack_events, args=(), daemon=True)
+        slack_thread.start()
+
+        timer_thread = Thread(target=self.periodically_update_slack, args=(), daemon=True)
+        timer_thread.start()
+
+    @abc.abstractmethod
+    def get_slack_client(self) -> SlackClient:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_slack_channel(self) -> str:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_deployment_name(self) -> str:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_progress(self) -> str:
+        raise NotImplementedError()
+
+    def get_active_button(self) -> Optional[str]:
+        return None
+
+    @abc.abstractmethod
+    def get_button_element(self, button, is_active):
+        raise NotImplementedError()
+
+    def get_slack_blocks_for_deployment(self) -> List[Dict]:
+        status = getattr(self, 'state', None) or 'Uninitialized'
+        deployment_name = self.get_deployment_name()
+        message = self.human_readable_status
+        progress = self.get_progress()
+        last_action = self.last_action
+
+        button_elements = self.get_button_elements()
+        blocks = [
+            {
+                "type": "section",
+                "text": {
+                    "type": "mrkdwn",
+                    "text": f"{deployment_name}",
+                },
+            },
+            {
+                "type": "section",
+                "text": {
+                    "type": "mrkdwn",
+                    "text": message,
+                },
+            },
+            {
+                "type": "section",
+                "text": {
+                    "type": "mrkdwn",
+                    "text": f"State machine: `{status}`\nProgress: {progress}\nLast operator action: {last_action}",
+                },
+            },
+        ]
+
+        slo_text = self.get_slo_text()
+        if slo_text:
+            blocks.append({
+                "type": "section",
+                "text": {
+                    "type": "mrkdwn",
+                    "text": self.get_slo_text(),
+                },
+            })
+
+        if button_elements != []:
+            blocks.append({
+                "type": "actions",
+                "block_id": "deployment_actions",
+                "elements": button_elements,
+            })
+        return blocks
+
+    def get_slo_text(self) -> str:
```

Comment:
Not sure if this belongs in SlackDeployProcess or not, especially since `slo_watchers` is just instantiated by MFD. 

Either the SlackDeployProcess knows about SLOs and has an abstract method like `get_slo_watchers`, or this method should be abstract and be named something more general like `get_extra_deploy_text` or something?

Reply:
Agreed, this does feel like the wrong place for it. I like the get_extra_deploy_text idea.